src/Storages/ExternalStream/Kafka/Kafka.cpp

```diff
@@ -466,10 +474,34 @@ Pipe Kafka::read(
         else
             header = storage_snapshot->getSampleBlockForColumns({ProtonConsts::RESERVED_EVENT_TIME});
 
-        auto offsets = getOffsets(*consumer, query_info.seek_to_info, shards_to_query);
+        auto seek_to_info = query_info.seek_to_info;
+        /// seek_to defaults to 'latest' for streaming. In non-streaming case, 'earliest' is preferred.
+        if (!streaming && seek_to_info->getSeekTo().empty())
+            seek_to_info = std::make_shared<SeekToInfo>("earliest");
+
+        auto offsets = getOffsets(*consumer, seek_to_info, shards_to_query);
         assert(offsets.size() == shards_to_query.size());
 
         for (auto [shard, offset] : std::ranges::views::zip(shards_to_query, offsets))
+        {
+            Int64 high_watermark = NO_WARTERMARK;
+            if (!streaming)
+            {
+                auto marks = topic_ptr->queryWatermarks(shard);
+                LOG_INFO(logger, "watermarks on parition {} low={} high={}", shard, marks.low, marks.high);
+                high_watermark = marks.high;
+
+                if (marks.low == marks.high) /// there are no messages in the topic
+                {
+                    /// As there are no messages, no need to create a KafkaSource instance at all.
+                    pipes.emplace_back(std::make_shared<NullSource>(header));
```

Comment:
Can we just `continue` without NullSource ?

Reply:
No. If we even remove NullSource, a simple `select * from table(ex_stream)` will report missing column error like this

```
<Error> executeQuery: Code: 8. DB::Exception: Cannot find column `message` in source stream, there are only columns: []. (THERE_IS_NO_COLUMN) (version 1.5.8) (from [::1]:44578) (in query: select * from table(foo_raw);)
```